### PR TITLE
[SPARK-29390][SQL] Add the justify_days(), justify_hours() and justif_interval() functions

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Add 3 interval functions justify_days, justify_hours, justif_interval to support justify interval values

### Why are the changes needed?

For feature parity with postgres

add three interval functions to justify interval values.

justify_days(interval) | interval | Adjust interval so 30-day time periods are represented as months | justify_days(interval '35 days') | 1 mon 5 days
-- | -- | -- | -- | --
justify_hours(interval) | interval | Adjust interval so 24-hour time periods are represented as days | justify_hours(interval '27 hours') | 1 day 03:00:00
justify_interval(interval) | interval | Adjust interval using justify_days and justify_hours, with additional sign adjustments | justify_interval(interval '1 mon -1 hour') | 29 days 23:00:00

### Does this PR introduce any user-facing change?

yes. new interval functions are added

### How was this patch tested?

add ut

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @MaxGekk @maropu @dongjoon-hyun @wangyum, thanks in advance

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala [hunk 1]
```diff
@@ -257,3 +257,69 @@ case class MakeInterval(
 
   override def prettyName: String = "make_interval"
 }
+
+abstract class IntervalJustifyLike(
+    child: Expression,
+    justify: CalendarInterval => CalendarInterval,
+    justifyType: String) extends UnaryExpression with ExpectsInputTypes {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala]: nit: it's clearer to pass in `justifyFuncName: String`
```
extends IntervalJustifyLike(child, justifyHours, "justifyHours")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 2]
```diff
@@ -637,4 +637,39 @@ object IntervalUtils {
 
     new CalendarInterval(totalMonths, totalDays, micros)
   }
+
+  /**
+   * Adjust interval so 30-day time periods are represented as months
+   */
+  def justifyDays(interval: CalendarInterval): CalendarInterval = {
+    val months = Math.addExact(interval.months, interval.days / DAYS_PER_MONTH.toInt)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: if `days` is negative, will we end like `-1 month 2 days`?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: good catch, we shall calculate all days first just like other 2 funcs

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql [hunk 3]
```diff
@@ -342,3 +333,11 @@ SELECT interval '1 2:03:04' minute to second;
 -- select make_interval(secs := 'inf');
 -- select make_interval(secs := 'NaN');
 -- select make_interval(secs := 7e12);
+
+-- test justify_hours() and justify_days()
+-- [SPARK-29390] Add the justify_days(), justify_hours() and justify_interval() functions
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql]: do we still need this?

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql]: I have some similar PRs which also modify PostgreSQL ported tests, mostly will be requested to leave or enable these tests. I guess they just want to track the progress of PostgreSQL feature parity in Spark.

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql]: I don't get it. This is a TODO and now it's done. Why still keep it?

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql]: I'd rather remove these too. I will push a new commit to delete these.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql [hunk 4]
```diff
@@ -156,15 +156,6 @@
 -- [SPARK-29389] Support synonyms for interval units
 -- select '100000000y 10mon -1000000000d -100000h -10min -10.000001s ago'::interval;
 
--- test justify_hours() and justify_days()
--- [SPARK-29390] Add the justify_days(), justify_hours() and justify_interval() functions
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql]: seems like there is a misunderstanding... I was asking to remove this TODO comment, not the tests.

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql]: my fault

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 5]
```diff
@@ -637,4 +637,40 @@ object IntervalUtils {
 
     new CalendarInterval(totalMonths, totalDays, micros)
   }
+
+  /**
+   * Adjust interval so 30-day time periods are represented as months
+   */
+  def justifyDays(interval: CalendarInterval): CalendarInterval = {
+    val monthToDays = interval.months * DAYS_PER_MONTH
+    val totalDays = Math.addExact(monthToDays, interval.days)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: can we overflow here? The worst case is `Int.Max * DAYS_PER_MONTH + Int.Max`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 6]
```diff
@@ -637,4 +637,40 @@ object IntervalUtils {
...
+    val months = Math.toIntExact(totalDays / DAYS_PER_MONTH)
+    val days = Math.toIntExact(totalDays % DAYS_PER_MONTH)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: ditto, the value is less than 30 and can never overflow.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 7]
```diff
@@ -637,4 +637,40 @@ object IntervalUtils {
...
+   * Adjust interval using justifyHours and justifyDays, with additional sign adjustments
+   */
+  def justifyInterval(interval: CalendarInterval): CalendarInterval = {
+    val monthToDays = DAYS_PER_MONTH * interval.months
+    val dayToUs = Math.multiplyExact(monthToDays + interval.days, MICROS_PER_DAY)
+    val totalUs = Math.addExact(interval.microseconds, dayToUs)
+    val microseconds = totalUs % MICROS_PER_DAY
+    val totalDays = totalUs / MICROS_PER_DAY
+    val days = totalDays % DAYS_PER_MONTH
+    val months = Math.toIntExact(totalDays / DAYS_PER_MONTH)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: The worse case is `Long.Max / MICROS_PER_DAY  / DAYS_PER_MONTH`, which can't overflow.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql [hunk 8]
```diff
@@ -342,3 +333,10 @@ SELECT interval '1 2:03:04' minute to second;
 -- select make_interval(secs := 'inf');
 -- select make_interval(secs := 'NaN');
 -- select make_interval(secs := 7e12);
+
+-- test justify_hours() and justify_days()
+SELECT justify_hours(interval '6 months 3 days 52 hours 3 minutes 2 seconds') as `6 mons 5 days 4 hours 3 mins 2 seconds`;
+SELECT justify_days(interval '6 months 36 days 5 hours 4 minutes 3 seconds') as `7 mons 6 days 5 hours 4 mins 3 seconds`;
+
+-- test justify_interval()
+SELECT justify_interval(interval '1 month -1 hour') as `1 month -1 hour`;
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql]: Can you move these tests into the original position? I just want to keep the diff between this test file and the Pg original test smaller.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 9]
```diff
@@ -637,4 +637,40 @@ object IntervalUtils {
 
     new CalendarInterval(totalMonths, totalDays, micros)
   }
+
+  /**
+   * Adjust interval so 30-day time periods are represented as months
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: super nit: add `.` in the end along with the other comments above?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala [hunk 10]
```diff
@@ -257,3 +257,69 @@ case class MakeInterval(
...
+    justifyFuncName: String) extends UnaryExpression with ExpectsInputTypes {
+  override def inputTypes: Seq[AbstractDataType] = Seq(CalendarIntervalType)
+
+  override def dataType: DataType = CalendarIntervalType
+
+  override def nullSafeEval(input: Any): Any = {
+    try {
+      justify(input.asInstanceOf[CalendarInterval])
+    } catch {
+      case _: ArithmeticException => null
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala]: nit: `case NonFatal(_) =>`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 7 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: ok, i'd remove these useless overflow check

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql (hunk 8 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/interval.sql]: OK, I just wanted to make the diff smalle. I'd move them back

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by gatorsmile** [comment]
Sorry for my late response. This PR introduces the 400+ LOC. Any system supports this? This looks very specific to pgSQL. Not sure whether we should just simply ignore this support and reduce the maintenance cost.

**Comment by maropu** [comment]
@yaooqinn Can you check if the other DBMS-like systems support them?

**Comment by yaooqinn** [comment]
This is PostgreSQL specific. But I think the meanings of these functions are unambiguous and they won't cost much for maintenance.

**Comment by gatorsmile** [comment]
I think we should still revert it. I do not see the value of this functions.

**Pull request closed without merging**